components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -1812,125 +1816,341 @@ private int handleStartIndexEqualsNULL(Integer startIndex) {
-                    "attribute value : " + attributeValue + " and filter operation " + filterOperation, e);
+            throw new CharonException(
+                    "Error in filtering groups by attribute name : " + attributeName + ", " + "attribute value : "
+                            + attributeValue + " and filter operation : " + filterOperation, e);
         } catch (org.wso2.carbon.user.api.UserStoreException e) {
-            throw new CharonException("Error in filtering group with filter: "
-                    + attributeName + filterOperation + attributeValue, e);
-        } catch (IdentitySCIMException e) {
-            throw new CharonException("Error in retrieving SCIM Group information from database.", e);
-        } catch (BadRequestException e) {
-            throw new CharonException("Error in retrieving SCIM Group.", e);
+            throw new CharonException(
+                    "Error in filtering group with filter: " + attributeName + " + " + filterOperation + " + "
+                            + attributeValue, e);
         }
-        //set the totalResults value in index 0
+        // Set the totalResults value in index 0.
         filteredGroups.set(0, filteredGroups.size() - 1);
         return filteredGroups;
     }
 
+    /**
+     * Resolve the domain name in request according to isFilterUsersAndGroupsOnlyFromPrimaryDomainEnabled() or
+     * isFilteringEnhancementsEnabled() or domain name embedded in the filter attribute value.
+     *
+     * @param domainName Domain name passed in the request.
+     * @param node       Expression node
+     * @return Domain name
+     * @throws CharonException
+     */
+    private String resolveDomain(String domainName, ExpressionNode node) throws CharonException {
+        try {
+            // Update the domain name if a domain is appended to the attribute value.
+            domainName = resolveDomainInAttributeValue(domainName, node);
+
+            // Apply filter enhancements if the domain is not specified in the request.
+            if (StringUtils.isEmpty(domainName)) {
+                domainName = applyFilterPropertiesForDomain(node);
+            }
+            return domainName;
+        } catch (BadRequestException e) {
+            String errorMessage = String
+                    .format(" Domain name in the attribute value: %s does not match with the domain parameter: %s in "
+                            + "the request.", node.getValue(), domainName);
+            throw new CharonException(errorMessage, e);
+        }
+    }
+
+    /**
+     * Method to check isFilterUsersAndGroupsOnlyFromPrimaryDomainEnabled() or isFilteringEnhancementsEnabled() which
+     * enables filtering in primary domain only.
+     *
+     * @param node Expression node.
+     * @return Primary domain name if properties are enabled or return NULL when properties are disabled.
+     */
+    private String applyFilterPropertiesForDomain(ExpressionNode node) {
```

Comment:
Shall we change the name of this method silmilar to `getDomainWithFilterProperties`? Also better to generalize javadoc description for the method, without specifically mentioning the property names

Reply:
+1 for changing the name. But it was decided to mention the property names in the last code review.